File: 695-max-area-of-island/max-area-of-island.py
class Solution:
    def DFS(self,grid,row,col,seen):

        stack = [(row,col)]
        seen.add((row,col))
        area = 1

        while stack:
            cand_row,cand_col = stack.pop()

            neighbs = [
                (cand_row -1,cand_col),
                (cand_row + 1,cand_col),
                (cand_row,cand_col -1),
                (cand_row,cand_col + 1)
            ]

            for neighb_row,neighb_col in neighbs:
                # check out of bounds, seen and is not water
                if 0 <= neighb_row < len(grid) and 0<= neighb_col < len(grid[0]) and (neighb_row,neighb_col) not in seen and grid[neighb_row][neighb_col] != 0:
                    stack.append((neighb_row,neighb_col))
                    seen.add((neighb_row,neighb_col))
                    area += 1

        return area

    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        seen = set()
        maxArea = 0
        for row in range(len(grid)):
            for col in range(len(grid[0])):
                if grid[row][col] == 1:
                    area = self.DFS(grid,row,col,seen) #DFS to find area of that island
                    maxArea = max(maxArea,area) # update max area

        return maxArea

File: 695-max-area-of-island/max-area-of-island.cpp
class Solution {
public:
    vector<int> dx = {-1,0,1,0}, dy = {0,1,0,-1};

    // DFS
    int DFS(int row, int col, vector<vector<int>>& grid)
    {
        grid[row][col] = 0;
        int area {1};

        for(int i {}; i< 4; i++)
        {
            int nr = row + dx[i];
            int nc = col + dy[i];

            // check error out of bounds, and if if neighb cell is land
            if(nr >= 0 && nr<grid.size() && 0<=nc && nc<grid[0].size() && grid[nr][nc] == 1)
            {
                area += DFS(nr,nc,grid); // recurse further 
            }
        }

        return area;
    }

    int maxAreaOfIsland(vector<vector<int>>& grid) {
        int result {};
        for(int i {}; i < grid.size(); i++)
        {
            for(int j {}; j < grid[0].size(); j++)
            {
                // run a DFS if we encounter an island
                if(grid[i][j] == 1)
                {
                    result = std::max(result, DFS(i,j,grid));
                }
            }
        }
        return result;
    }
};